internal/blocksync/reactor.go

```diff
@@ -185,31 +185,39 @@ func (r *Reactor) OnStop() {
 // Otherwise, we'll respond saying we do not have it.
 func (r *Reactor) respondToPeer(ctx context.Context, msg *bcproto.BlockRequest, peerID types.NodeID, blockSyncCh *p2p.Channel) error {
 	block := r.store.LoadBlock(msg.Height)
-	if block != nil {
-		extCommit := r.store.LoadBlockExtendedCommit(msg.Height)
-		if extCommit == nil {
-			return fmt.Errorf("found block in store without extended commit: %v", block)
-		}
-		blockProto, err := block.ToProto()
-		if err != nil {
-			return fmt.Errorf("failed to convert block to protobuf: %w", err)
-		}
-
+	if block == nil {
+		r.logger.Info("peer requesting a block we do not have", "peer", peerID, "height", msg.Height)
 		return blockSyncCh.Send(ctx, p2p.Envelope{
-			To: peerID,
-			Message: &bcproto.BlockResponse{
-				Block:     blockProto,
-				ExtCommit: extCommit.ToProto(),
-			},
+			To:      peerID,
+			Message: &bcproto.NoBlockResponse{Height: msg.Height},
 		})
 	}
 
-	r.logger.Info("peer requesting a block we do not have", "peer", peerID, "height", msg.Height)
+	state, err := r.stateStore.Load()
+	if err != nil {
+		return fmt.Errorf("loading state: %w", err)
+	}
+	var extCommit *types.ExtendedCommit
+	if state.ConsensusParams.ABCI.VoteExtensionsEnabled(msg.Height) {
+		extCommit = r.store.LoadBlockExtendedCommit(msg.Height)
+		if extCommit == nil {
+			return fmt.Errorf("found block in store with no extended commit: %v", block)
+		}
+	}
+
+	blockProto, err := block.ToProto()
+	if err != nil {
+		return fmt.Errorf("failed to convert block to protobuf: %w", err)
+	}
 
 	return blockSyncCh.Send(ctx, p2p.Envelope{
-		To:      peerID,
-		Message: &bcproto.NoBlockResponse{Height: msg.Height},
+		To: peerID,
+		Message: &bcproto.BlockResponse{
+			Block:     blockProto,
+			ExtCommit: extCommit.ToProto(),
```

Comment:
Very minor. Since we're dealing with errors in `block.ToProto()`, do we also want to deal with `extCommit.ToProto()`? (other then `extCommit` being `nil`)

Reply:
`block.ToProto()` returns an error, `extCommit.ToProto()` does not. This would not compile if `extCommit.ToProto()` returned an error.

Reply:
TIL ... :-)

Reply:
I know Go gets a lot of hate for the `if err != nil{..}` idiom, but I do think it enforces some level of discipline around making sure you actually check and handle/propagate the errors from the functions you call.

Reply:
> making sure you actually check and handle/propagate the errors from the functions you call.

That's precisely what I hate the most in C, C++, and the rest of the gang.
That, and coercing `int`s into booleans.